### PR TITLE
UCP/RMA: Fix PUT-SGL protocol selection length

### PR DESCRIPTION
**Pull request by usberkeley**
## What?
Use the total SGL byte length for PUT-SGL protocol selection and reconfiguration. Keep the CUDA SGL tests skipped for now.

## Why?
The SGL iterator length is an element count, but protocol selection treated it as bytes. For example, four 2 KiB entries were selected as a 4-byte message instead of 8 KiB. This could lead to incorrect protocol ranges and reconfiguration decisions.

The sporadic CUDA failure reproduced in this PR has a separate root cause in `cuda_copy` reachability and is
outside the scope of this fix.

## How?
Calculate the total byte length once, with overflow checking, and store it in the existing request length field. Use that value for initial selection, reconfiguration, and error reporting. Add focused regression tests for the byte length and overflow cases.

Leave the CUDA SGL tests skipped. They will be re-enabled in a separate change that fixes the `cuda_copy` reachability issue.

### CODE REVIEW

File: src/ucp/dt/dt_sgl.c [hunk 1]
```diff
@@ -15,6 +15,25 @@
...
 
 
+ucs_status_t ucp_dt_sgl_get_length(const size_t *lengths, size_t count,
+                                   size_t *total_length_p)
+{
+    size_t total_length = 0;
+    size_t i;
+
+    for (i = 0; i < count; ++i) {
+        if (ucs_unlikely(lengths[i] > (SIZE_MAX - total_length))) {
```
Comment (michal-shalev) [inline src/ucp/dt/dt_sgl.c]: Is this check needed? `ucp_dt_iov_length()` sums the IOV lengths without it, and here the total only feeds protocol selection, and the per-element branch prevents the loop from vectorizing.

File: src/ucp/dt/dt_sgl.c [hunk 2]
```diff
@@ -15,6 +15,25 @@
 #include <ucp/core/ucp_rkey.h>
 
 
+ucs_status_t ucp_dt_sgl_get_length(const size_t *lengths, size_t count,
```
Comment (michal-shalev) [inline src/ucp/dt/dt_sgl.c]: Could this be a `static inline` in `dt_sgl.h` returning `size_t`, like `ucp_dt_iov_length()` in `dt_iov.h`, and be called from the SGL branch of `ucp_datatype_iter_init()`? The other functions in `dt_sgl.c` are out of line, but they only run under `ENABLE_PARAMS_CHECK`, while this one is on the release fast path, so it ends up as a real call from an always-inline function.

File: src/ucp/proto/proto_common.inl [hunk 3]
```diff
@@ -313,11 +322,21 @@ ucp_proto_request_send_op(ucp_ep_h ep, ucp_proto_select_t *proto_select,
         return UCS_STATUS_PTR(status);
     }
 
+    if (req->send.state.dt_iter.dt_class == UCP_DATATYPE_SGL) {
```
Comment (michal-shalev) [inline src/ucp/proto/proto_common.inl]: Could this move into the SGL branch inside `ucp_datatype_iter_init()`, similar to how IOV computes its own total via `ucp_dt_iov_length()`? That branch is taken anyway, while this check runs on every tag, AM and stream send too.

File: src/ucp/proto/proto_common.inl [hunk 4]
```diff
@@ -11,9 +11,18 @@
 #include "proto_select.inl"
 
 #include <ucp/dt/datatype_iter.inl>
+#include <ucp/dt/dt_sgl.h>
 #include <ucp/core/ucp_request.inl>
 
 
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_request_data_length(const ucp_request_t *req)
```
Comment (michal-shalev) [inline src/ucp/proto/proto_common.inl]: Can we avoid making the proto layer aware of SGL internals and remove the accessor altogether by keeping `dt_iter->length` in bytes for SGL too, with the element cursor in `type.sgl` like IOV keeps `iov_index/iov_offset`?

File: test/gtest/ucp/test_ucp_rma.cc [hunk 5]
```diff
@@ -1179,6 +1190,16 @@ class test_ucp_rma_sgl : public test_ucp_rma {
...
                                            UCP_REMOTE_ADDR_INVALID,
                                            UCP_RKEY_INVALID, &param);
+        if (check_request_length) {
+            size_t expected_length = 0;
+
+            for (size_t elem_size : elem_sizes) {
+                expected_length += elem_size;
+            }
+
+            EXPECT_EQ(expected_length, req->send.length);
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_rma.cc]: Could this verify protocol selection rather than `req->send.length` itself? As written it passes even if selection keeps using `dt_iter.length`.

File: src/ucp/dt/dt_sgl.c (hunk 1 above)
Comment (usberkeley) [reply src/ucp/dt/dt_sgl.c]: Got it, I’ve remove unnecessary check.

File: src/ucp/dt/dt_sgl.c (hunk 2 above)
Comment (usberkeley) [reply src/ucp/dt/dt_sgl.c]: My mistake, i didn’t catch this. Got it - Done

File: src/ucp/proto/proto_common.inl (hunk 4 above)
Comment (usberkeley) [reply src/ucp/proto/proto_common.inl]: Yeah, this is definitely better than my original design. It keeps the field semantics consistent and avoids ambiguity. Thanks.

File: src/ucp/proto/proto_common.inl (hunk 3 above)
Comment (usberkeley) [reply src/ucp/proto/proto_common.inl]: Got it, done.

File: test/gtest/ucp/test_ucp_rma.cc (hunk 5 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_rma.cc]: My mistake, i didn’t catch this. Got it - Done

**Comment by brminich** [comment]
@usberkeley thanks for the patch! Do you have CLA signed?

**Comment by usberkeley** [comment]
> @usberkeley thanks for the patch! Do you have CLA signed?

Hi @brminich
I’m not sure. How can I check whether I’ve signed it? This is my first PR for UCX

Just to clarify, I’ve added my name to the `AUTHORS` file – I’m not sure if that’s what you’re referring to.

File: src/ucp/proto/proto_common.inl (hunk 4 above)
Comment (usberkeley) [reply src/ucp/proto/proto_common.inl]: Hi @michal-shalev 

I tried exactly this: `dt_iter->length/offset` are now byte-based, while `type.sgl.count/index` track elements. However, adding the two `size_t` fields grows `ucp_request_t` from 280 to 296 bytes and fails `test_obj_size`. 

`ucp_request_t` is a core object on the UCX communication hot path. I’m not sure whether increasing the `test_obj_size` limit from 280 to 296 bytes would be acceptable, since it may affect cache efficiency, memory-pool utilization, and overall performance.

Comment (michal-shalev) [reply src/ucp/proto/proto_common.inl]: IMO bumping it here is fine, since the increase is temporary.

Have a look at #11647 (8bcbe19), I moved `remote_addrs`/`rkeys` out of the iterator into `ucp_request_t` there, which takes two pointers out of `type.sgl`. So once that lands the net result of both PRs is 280, same as today. No need to restructure anything on your side.

File: src/ucp/dt/datatype_iter.h [hunk 6]
```diff
@@ -44,8 +44,8 @@ typedef struct {
     ucp_dt_class_t    dt_class; /* Datatype class (contig/iov/...) */
     ucp_memory_info_t mem_info; /* Memory type and locality, needed to
                                    pack/unpack */
-    size_t            length; /* Total packed flat length */
-    size_t            offset; /* Current flat offset */
+    size_t            length; /* Packed byte length */
+    size_t            offset; /* Current byte offset */
```
Comment (michal-shalev) [inline src/ucp/dt/datatype_iter.h]: minor: can we keep the original comments?
"flat" seems to be the existing term for this in UCX (`ucs_iov_iter_flat_offset`, and the iov TODO below still says "flat length").

File: src/ucp/dt/datatype_iter.h [hunk 7]
```diff
@@ -79,7 +79,8 @@ typedef struct {
             ucp_mem_h        *memhs;
             const uint64_t   *remote_addrs;
             ucp_rkey_h const *rkeys;
-            /* length = element count, offset = current element index */
+            size_t            count;
+            size_t            index;
```
Comment (michal-shalev) [inline src/ucp/dt/datatype_iter.h]:
```suggestion
            size_t           elem_count;
            size_t           elem_index;
```
minor: align them with the `*` column like the other union arms, and rename to match `max_elem_count` in `next_sgl()`.

**Comment by michal-shalev** [comment]
> > @usberkeley thanks for the patch! Do you have CLA signed?
> 
> Hi @brminich I’m not sure. How can I check whether I’ve signed it? This is my first PR for UCX
> 
> Just to clarify, I’ve added my name to the `AUTHORS` file – I’m not sure if that’s what you’re referring to.

According to the [Guidance for contributors](https://github.com/openucx/ucx/wiki/Guidance-for-contributors): 

> UCX uses Git for source code management, Azure CI for continuous integration, and Google’s C++ test framework for unit testing. Before you make any contributions to the projects, please sign a copy of our CLA https://openucx.org/license/ and email the CLA to admin at ucfconsortium.org

you need to sign the UCX CLA and send it to admin@ucfconsortium.org.

File: src/ucp/proto/proto_common.inl (hunk 4 above)
Comment (usberkeley) [reply src/ucp/proto/proto_common.inl]: Got it

File: src/ucp/dt/datatype_iter.h (hunk 6 above)
Comment (usberkeley) [reply src/ucp/dt/datatype_iter.h]: Sure, no problem - Done

**Comment by usberkeley** [comment]
> > > @usberkeley thanks for the patch! Do you have CLA signed?
> > 
> > 
> > Hi @brminich I’m not sure. How can I check whether I’ve signed it? This is my first PR for UCX
> > Just to clarify, I’ve added my name to the `AUTHORS` file – I’m not sure if that’s what you’re referring to.
> 
> According to the [Guidance for contributors](https://github.com/openucx/ucx/wiki/Guidance-for-contributors):
> 
> > UCX uses Git for source code management, Azure CI for continuous integration, and Google’s C++ test framework for unit testing. Before you make any contributions to the projects, please sign a copy of our CLA https://openucx.org/license/ and email the CLA to admin at ucfconsortium.org
> 
> you need to sign the UCX CLA and send it to [admin@ucfconsortium.org](mailto:admin@ucfconsortium.org).

Thanks for the detailed guidance — Done, I’ve sent the CLA email

File: src/ucp/dt/datatype_iter.h [hunk 8]
```diff
@@ -79,7 +79,9 @@ typedef struct {
             ucp_mem_h        *memhs;
             const uint64_t   *remote_addrs;
             ucp_rkey_h const *rkeys;
-            /* length = element count, offset = current element index */
+            size_t            count;
```
Comment (michal-shalev) [inline src/ucp/dt/datatype_iter.h]: This should have been renamed to `elem_count`, not kept next to it.
libucp doesn't compile now, since `type.sgl.index` no longer exists but is still used in `datatype_iter.c`, `datatype_iter.inl`, `put_offload.c` and `test_ucp_dt.cc`.

File: src/ucp/dt/datatype_iter.h [hunk 9]
```diff
@@ -79,7 +79,9 @@ typedef struct {
...
+            size_t           elem_count;
+            size_t           elem_index;
```
Comment (michal-shalev) [inline src/ucp/dt/datatype_iter.h]: Once it's two fields again, `EXPECTED_SIZE(ucp_request_t, 280)` in `test/gtest/common/test_obj_size.cc` needs to become 296 for CI to pass.

File: src/ucp/dt/datatype_iter.inl [hunk 10]
```diff
@@ -623,8 +627,20 @@ ucp_datatype_iter_next_sgl(const ucp_datatype_iter_t *dt_iter,
                            size_t max_elem_count,
                            ucp_datatype_iter_t *next_iter)
 {
+    size_t index, elem_count;
```
Comment (michal-shalev) [inline src/ucp/dt/datatype_iter.inl]: Minor: maybe rename to `size_t elem_index, elem_count;`?

File: src/ucp/dt/datatype_iter.h (hunk 8 above)
Comment (usberkeley) [reply src/ucp/dt/datatype_iter.h]: My bad for the repeated reviews. I'll step it up next time ❤️

File: src/ucp/dt/datatype_iter.h (hunk 9 above)
Comment (usberkeley) [reply src/ucp/dt/datatype_iter.h]: Got it -- Done

File: src/ucp/dt/datatype_iter.inl (hunk 10 above)
Comment (usberkeley) [reply src/ucp/dt/datatype_iter.inl]: Got it -- Done

File: src/ucp/proto/proto_reconfig.c [hunk 11]
```diff
@@ -101,8 +101,9 @@ static ucs_status_t ucp_proto_reconfig_progress(uct_pending_req_t *self)
                                   req->send.proto_config->rkey_cfg_index,
                                   &req->send.proto_config->select_param,
                                   ucp_operation_names, &strb);
-        ucs_error("cannot find remote protocol for: %s",
-                  ucs_string_buffer_cstr(&strb));
+        ucs_error("cannot find remote protocol for: %s, %zu bytes",
+                  ucs_string_buffer_cstr(&strb),
+                  req->send.state.dt_iter.length);
```
Comment (michal-shalev) [inline src/ucp/proto/proto_reconfig.c]: Is this change needed for the fix?

File: test/gtest/ucp/test_ucp_rma.cc [hunk 12]
```diff
@@ -1154,11 +1169,24 @@ class test_ucp_rma_sgl : public test_ucp_rma {
...
 
+        if (check_proto_selection) {
+            ucp_context_attr_t attr = {};
+
+            attr.field_mask = UCP_ATTR_FIELD_REQUEST_SIZE;
+            ASSERT_UCS_OK(ucp_context_query(sender().ucph(), &attr));
+            request_mem.reset(new uint8_t[attr.request_size + 1]);
+            param.op_attr_mask |= UCP_OP_ATTR_FIELD_REQUEST;
+            param.request    = request_mem.get() + attr.request_size;
+            req              = static_cast<ucp_request_t*>(param.request) - 1;
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_rma.cc]: Minor: could you align these three assignments to the same column?

**Comment by usberkeley** [comment]
Hi @michal-shalev 
I found that the incorrect data size calculation is not the cause of the “CUDA SGL transport protocol not found” error.

This PR will first fix the data size issue. Would it make sense to re-enable the skipped CUDA test first, and fix the actual root cause in a separate PR?

#### Root cause analysis

The test uses two UCP workers in the same process. cuda_ipc rejects same-process communication, so it should fall back to cuda_copy.

However, each cuda_copy iface gets a different random ID (cuda_copy_iface.c:341), while connection setup requires the IDs to match (cuda_copy_iface.c:89). As a result, the two workers are incorrectly considered unreachable.

Also, although the CI driver supports CUDA 13.0, UCX is built with CUDA 12.8. CUDA IPC only advertises native SGL support when built with CUDA >= 13.0.

**Comment by michal-shalev** [comment]
> Hi @michal-shalev I found that the incorrect data size calculation is not the cause of the “CUDA SGL transport protocol not found” error.
> 
> This PR will first fix the data size issue. Would it make sense to re-enable the skipped CUDA test first, and fix the actual root cause in a separate PR?
> 
> #### Root cause analysis
> The test uses two UCP workers in the same process. cuda_ipc rejects same-process communication, so it should fall back to cuda_copy.
> 
> However, each cuda_copy iface gets a different random ID (cuda_copy_iface.c:341), while connection setup requires the IDs to match (cuda_copy_iface.c:89). As a result, the two workers are incorrectly considered unreachable.
> 
> Also, although the CI driver supports CUDA 13.0, UCX is built with CUDA 12.8. CUDA IPC only advertises native SGL support when built with CUDA >= 13.0.

Thanks for digging into this. I can indeed see that the sporadic CUDA failure is still reproduced in this PR.

Since the CUDA failure has a separate root cause, I think this PR should stay focused on fixing the SGL byte-length semantics, while keeping the CUDA test skipped for now.

Then the `cuda_copy` reachability issue can be fixed separately, and the CUDA test can be re-enabled as part of that fix.

Could you also update the PR description accordingly?

**Comment by usberkeley** [comment]
> > Hi @michal-shalev I found that the incorrect data size calculation is not the cause of the “CUDA SGL transport protocol not found” error.
> > This PR will first fix the data size issue. Would it make sense to re-enable the skipped CUDA test first, and fix the actual root cause in a separate PR?
> > #### Root cause analysis
> > The test uses two UCP workers in the same process. cuda_ipc rejects same-process communication, so it should fall back to cuda_copy.
> > However, each cuda_copy iface gets a different random ID (cuda_copy_iface.c:341), while connection setup requires the IDs to match (cuda_copy_iface.c:89). As a result, the two workers are incorrectly considered unreachable.
> > Also, although the CI driver supports CUDA 13.0, UCX is built with CUDA 12.8. CUDA IPC only advertises native SGL support when built with CUDA >= 13.0.
> 
> Thanks for digging into this. I can indeed see that the sporadic CUDA failure is still reproduced in this PR.
> 
> Since the CUDA failure has a separate root cause, I think this PR should stay focused on fixing the SGL byte-length semantics, while keeping the CUDA test skipped for now.
> 
> Then the `cuda_copy` reachability issue can be fixed separately, and the CUDA test can be re-enabled as part of that fix.
> 
> Could you also update the PR description accordingly?

Got it, no problem -- Done ❤️

File: test/gtest/ucp/test_ucp_rma.cc [hunk 13]
```diff
@@ -974,11 +975,6 @@ class test_ucp_rma_sgl : public test_ucp_rma {
     }
 
     virtual void init() override {
-        /* FIXME: sporadic failure on CUDA memory type. re-enable once fixed */
-        if (mem_type() == UCS_MEMORY_TYPE_CUDA) {
-            UCS_TEST_SKIP_R("sporadic failure on CUDA memory type");
-        }
-
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_rma.cc:1020]: Since the CUDA failure has a different root cause, could we keep this skip here and re-enable it in the PR that fixes it?

Comment (usberkeley) [reply test/gtest/ucp/test_ucp_rma.cc:1020]: Got it -- Done

File: test/gtest/ucp/test_ucp_rma.cc [hunk 14]
```diff
@@ -1154,11 +1169,24 @@ class test_ucp_rma_sgl : public test_ucp_rma {
...
+        if (check_proto_selection) {
+            ucp_context_attr_t attr = {};
+
+            attr.field_mask = UCP_ATTR_FIELD_REQUEST_SIZE;
+            ASSERT_UCS_OK(ucp_context_query(sender().ucph(), &attr));
+            request_mem.reset(new uint8_t[attr.request_size + 1]);
+            param.op_attr_mask |= UCP_OP_ATTR_FIELD_REQUEST;
+            param.request    = request_mem.get() + attr.request_size;
+            req              = static_cast<ucp_request_t*>(param.request) - 1;
+        }
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_rma.cc]: Could we drop the external request memory and use `(ucp_request_t*)sptr - 1` instead? UCX returns `req + 1` for internal requests too (`proto_common.inl:287`), and `test_ucp_request.cc:467` does exactly this, it would also let `ucp_request_release(sptr)` stay unconditional.

File: test/gtest/ucp/test_ucp_rma.cc (hunk 12 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_rma.cc]: Got it -- Done

File: src/ucp/proto/proto_reconfig.c (hunk 11 above)
Comment (usberkeley) [reply src/ucp/proto/proto_reconfig.c]: No needed, just to enhance logs -- Remove, done ❤️

File: test/gtest/ucp/test_ucp_dt.cc [hunk 15]
```diff
@@ -401,13 +405,29 @@ UCS_TEST_F(test_ucp_dt_sgl, iter_next_one_by_one) {
...
+
+    init_sgl_iter(3);
+    EXPECT_EQ(2u, ucp_datatype_iter_next_sgl(&m_dt_iter, 2, &next_iter));
+    ucp_datatype_iter_copy_position(&m_dt_iter, &next_iter, UINT_MAX);
+    EXPECT_NE(0u, m_dt_iter.offset);
+    EXPECT_EQ(2u, m_dt_iter.type.sgl.elem_index);
+
+    ucp_datatype_iter_rewind(&m_dt_iter, UINT_MAX);
+    EXPECT_EQ(0u, m_dt_iter.offset);
+    EXPECT_EQ(0u, m_dt_iter.type.sgl.elem_index);
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_dt.cc]: Minor: consider `ASSERT_*`

File: test/gtest/ucp/test_ucp_dt.cc [hunk 16]
```diff
@@ -401,13 +405,29 @@ UCS_TEST_F(test_ucp_dt_sgl, iter_next_one_by_one) {
...
     EXPECT_TRUE(ucp_datatype_iter_is_end(&m_dt_iter));
 }
 
+UCS_TEST_F(test_ucp_dt_sgl, iter_rewind) {
+    ucp_datatype_iter_t next_iter = {};
+
+    init_sgl_iter(3);
+    EXPECT_EQ(2u, ucp_datatype_iter_next_sgl(&m_dt_iter, 2, &next_iter));
+    ucp_datatype_iter_copy_position(&m_dt_iter, &next_iter, UINT_MAX);
+    EXPECT_NE(0u, m_dt_iter.offset);
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_dt.cc]: Should this check the exact byte offset? Would `EXPECT_NE(0u, ...)` pass with the old element-based value too?

File: test/gtest/ucp/test_ucp_rma.cc (hunk 14 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_rma.cc]: Got it, UCX allocates the request normally and inspecting its internal state via sptr - 1 is indeed much simpler now. ❤️

File: test/gtest/ucp/test_ucp_dt.cc (hunk 15 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_dt.cc]: Got it -- done

File: test/gtest/ucp/test_ucp_dt.cc (hunk 16 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_dt.cc]: Got it -- Done

File: test/gtest/ucp/test_ucp_rma.cc [hunk 17]
```diff
@@ -1218,6 +1249,7 @@ class test_ucp_rma_sgl : public test_ucp_rma {
         }
 
         ucp_request_release(sptr);
+
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_rma.cc:1305]: Minor: please remove this extra new line.

**Comment by michal-shalev** [comment]
@usberkeley can you please merge master?

File: test/gtest/ucp/test_ucp_rma.cc (hunk 17 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_rma.cc:1305]: Got it -- Done

**Comment by usberkeley** [comment]
> @usberkeley can you please merge master?

Got it --Done

File: test/gtest/ucp/test_ucp_rma.cc (hunk 17 above)
Comment (michal-shalev) [reply test/gtest/ucp/test_ucp_rma.cc:1305]: It is still there, can you please remove it?

File: test/gtest/ucp/test_ucp_rma.cc [hunk 18]
```diff
@@ -1216,8 +1247,8 @@ class test_ucp_rma_sgl : public test_ucp_rma {
                 ucp_worker_progress(receiver().worker());
             }
         }
-
```
Comment (michal-shalev) [inline test/gtest/ucp/test_ucp_rma.cc:1274]: Minor: no need to remove this line.

**Comment by michal-shalev** [comment]
> > @usberkeley can you please merge master?
> 
> Got it --Done

"This branch is out-of-date with the base branch"
Please merge upstream master.

File: test/gtest/ucp/test_ucp_rma.cc (hunk 17 above)
Comment (usberkeley) [reply test/gtest/ucp/test_ucp_rma.cc:1305]: My bad — could you check again?

**Comment by usberkeley** [comment]
> > > @usberkeley can you please merge master?
> > 
> > 
> > Got it --Done
> 
> "This branch is out-of-date with the base branch" Please merge upstream master.

Got it -- I’ll keep that in mind next time

**Review by michal-shalev — DISMISSED** [review]
LGTM. Thanks for contributing to UCX!

@brminich, can we merge this, or do you have any other comments on your end?

File: src/ucp/proto/proto_reconfig.c [hunk 19]
```diff
@@ -101,8 +101,6 @@ static ucs_status_t ucp_proto_reconfig_progress(uct_pending_req_t *self)
                                   req->send.proto_config->rkey_cfg_index,
                                   &req->send.proto_config->select_param,
                                   ucp_operation_names, &strb);
```
Comment (svc-nvidia-pr-review) [inline src/ucp/proto/proto_reconfig.c:103]: this log removal looks unrelated to the SGL length change — why drop it? this branch aborts the request and fails the EP, so the error message is useful for diagnosing the no-protocol case. if it's intentional, pls move it to a separate PR.

**Comment by usberkeley** [comment]
> LGTM. Thanks for contributing to UCX!
> 
> @brminich, can we merge this, or do you have any other comments on your end?

Hi @brminich 

When you have a chance, would you mind taking a look?

Also, we found that we can't send CLA emails to `admin@ucfconsortium.org` – could you also take a look? 
I tried a few times, and I also sent one to `info@openucx.org` – same issue.
```
[admin@ucfconsortium.org](mailto:admin@ucfconsortium.org)
d186762a.ess.barracudanetworks.com. (209.222.82.253, the server for the domain ucfconsortium.org.)
Remote server returned '554 4.7.0 <d186762a.ess.barracudanetworks.com. (209.222.82.253, the server for the domain ucfconsortium.org.) #4.7.0 smtp; 451 domain not found (not relaying for 209.85.216.48)>'

原始邮件头:

DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20251104; t=1787344322; x=1787949122; darn=ucfconsortium.org;
        h=mime-version:content-type:accept-language:message-id:date
         :thread-index:thread-topic:subject:to:from:from:to:cc:subject:date
         :message-id:reply-to:content-type;
        bh=6o8mEqKlVdqyN0vRPFN3p2KspyUUZEMKlnwA7y9LB2o=;
        b=lGP57yHpJlEqQnEnbLN2P3srXBzmlZUbR6BKtF9K60IwDEwRES7VUC2wkImJpX+nCJ
         G4CmDgv9PuBswLtHxF7HEkXz29Tnc/XQArdl0/iQ0kYIKgIEPrDlYHBVWkQFCCL+KHBZ
         Zi7mPVkXA3YBA0xqPpjad2ySGxD/adqVsl+iz3PPSFxkwo0wDIXpRFOjsOpYZ6pnzrIp
         vTdhoZtLEGtjaEQRoJ57K+SJUMraR1WcC1EWTp/qYF/Fc5e63R3w4N67GnvQ3W8KLeq2
         VVZhJPzI5v6fnQw2W2WxA63eolOV+T81y/pMJNsutd/9umCju8FOOUWUNrEVj4FS5KDD
         2/lQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20251104; t=1787344322; x=1787949122;
        h=mime-version:content-type:accept-language:message-id:date
         :thread-index:thread-topic:subject:to:from:x-gm-gg
         :x-gm-message-state:from:to:cc:subject:date:message-id:reply-to
         :content-type;
        bh=6o8mEqKlVdqyN0vRPFN3p2KspyUUZEMKlnwA7y9LB2o=;
        b=eX7hUGwIFy/PWQWFq19rC+YLM9W2YcASrAvjwn2+13pSpJPyI3I7796sQKogI94RCk
         tTgn7dRukhf8sIO0kD2oNLXNCgQJvPT4EUsRmaYkraEw1bdK7lU1ePiKb7tBxvAWrooJ
         gnbDeo2byX5aHbGT84ZerB5TlnW4MA/ejJ99JjlcGsMHsYEZonoFS2TyBNRIb4ae6Xph
         ss8gTWX7AwEPqUT9y+YFzLvONRFd4IhPRRY3pv/S8ci+B+hhPU9840NZSA55eBrpAXti
         UHENVSJC2ZITRG/acacIvckShT1xQV2D/v3tis45ho5oOIUgBeKc4bWULY5Lk0mJgq/C
         dNfA==
X-Gm-Message-State: AFuF++kPIhyS3rTgj9tAohLNzGJzM1vH9m9uGlIsoVWS6eF4wAJGOAJT
	hxCKP5QT2B4g7h4L0K3yPHFo5rw9jaFV6YptTsAK7Jgm8G3xqJlCLyxCXZY93A==
X-Gm-Gg: AR+sD11iKcSk4kgeUjmOhJnnNWUiajNHQ8PjfrIg6AW0/5Nu1vlSf1BzvDjv520sLXZ
	gHfgpLHV8uq4fZvQSHhf5nGA9UeIAJQgJuH8jhZ77hRkzJL1BZLwC6xqT83nBm68rgSrplRsH67
	vhN1tCvfV2LS+2W2VBXI5suuWGMuIT6tMu2otvBlXlfP0RsDXXnkgrR7v8tP0Rp8KJdc7gnCHSZ
	h3f53cVhCowuxB65jvwiHtg/e7RUkTUpbCje3zne1H9H5zLiuM7qY2T9FkjuOGqcpMhKczhhDp1
	2sr9unOKlIJGehx8Noop0VumMT5044q2/4FU1oFdujOJkoM7u9EZz0AWyTXqZ+uYT/ycEj7SYTT
	Ex0vJqWWgVd3KAJUBTii66uG/DKjIa5WtTeg9fhD20yzyENJwSa6CTJp5Se0FKoK0zzxSqvSTxx
	YvzcUaNOMhUibu8VmyNf0gCicduU2nn+1WBeAexHuwcLjsbCiKf06K9wilgGRUyGg5a0ucqvGAp
	wrn0jw4i3/nxQXgfe19Ep+epQbGNu9OlUZFMGK7z0zQkZo57XlNfxJsU1b+IvrVqQxF
X-Received: by 2002:a05:6a00:3916:b0:84e:2bb:2d3e with SMTP id d2e1a72fcca58-84fddfcd81amr37781292b3a.9.1787073285025;
        Tue, 18 Aug 2026 10:14:45 -0700 (PDT)
Return-Path: <[bradley.b.pitt@gmail.com](mailto:bradley.b.pitt@gmail.com)>
Received: from SEZPR04MB5651.apcprd04.prod.outlook.com ([2603:1046:101:38::5])
        by smtp.gmail.com with ESMTPSA id 41be03b00d2f7-cc13bf05371sm2429367a12.7.2026.08.18.10.14.42
        for <[admin@ucfconsortium.org](mailto:admin@ucfconsortium.org)>
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 18 Aug 2026 10:14:43 -0700 (PDT)
From: Bradley Pitt <[bradley.b.pitt@gmail.com](mailto:bradley.b.pitt@gmail.com)>
To: "[admin@ucfconsortium.org](mailto:admin@ucfconsortium.org)" <[admin@ucfconsortium.org](mailto:admin@ucfconsortium.org)>
Subject: UCX CLA
Thread-Topic: UCX CLA
Thread-Index: AQHdLzTwPMcSqMd4mU6+5bNp3nr3ig==
X-MS-Exchange-MessageSentRepresentingType: 1
Date: Tue, 18 Aug 2026 17:14:41 +0000
Message-ID: <[SEZPR04MB5651C4A64714BFAD27A0E0BBF8A62@SEZPR04MB5651.apcprd04.prod.outlook.com](mailto:SEZPR04MB5651C4A64714BFAD27A0E0BBF8A62@SEZPR04MB5651.apcprd04.prod.outlook.com)>
Accept-Language: zh-CN, en-US
X-MS-Has-Attach: yes
X-MS-Exchange-Organization-SCL: -1
X-MS-TNEF-Correlator:
X-MS-Exchange-Organization-RecordReviewCfmType: 0
x-ms-reactions: allow
Content-Type: text/plain
MIME-Version: 1.0
```

File: src/ucp/proto/proto_reconfig.c (hunk 19 above)
Comment (svc-nvidia-pr-review) [inline src/ucp/proto/proto_reconfig.c:103]: why is the `cannot find remote protocol for` error removed here? it looks unrelated to the SGL length fix, and it drops the diagnostic that reported `strb`. can this go in a separate PR (or be kept)?

**Comment by usberkeley** [comment]
Hi @michal-shalev 
This PR has quite a few conflicts with master, and I’ve force-pushed it a few times. Would it be okay if I just open a fresh PR?

I should have it ready in the next couple of days. I’ll clean things up and double-check everything before submitting.
